File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3003_maximize-the-number-of-partitions-after-operations/solutions/java/solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int maxPartitionsAfterOperations(String s, int k) {
        if (k == 0) return s.length();

        int n = s.length();
        int maxPartitions = 0;

        // Helper function to count maximum partitions for a given string
        java.util.function.ToIntFunction<String> getMaxPartitions = (str) -> {
            int count = 0;
            int start = 0;
            while (start < str.length()) {
                Map<Character, Integer> freq = new HashMap<>();
                int end = start;
                while (end < str.length() && freq.size() <= k) {
                    char ch = str.charAt(end);
                    freq.put(ch, freq.getOrDefault(ch, 0) + 1);
                    if (freq.size() > k) {
                        break;
                    }
                    end++;
                }
                count++;
                start = end;
            }
            return count;
        };

        // First check partitions without any change
        maxPartitions = getMaxPartitions.applyAsInt(s);

        // If k is 1, there is no point in changing a character to any other character
        if (k > 1) {
            // Try changing each character to each of 26 possible characters and calculate max partitions
            for (int i = 0; i < n; i++) {
                char originalChar = s.charAt(i);
                for (char newChar = 'a'; newChar <= 'z'; newChar++) {
                    if (newChar != originalChar) {
                        // Change s[i] to newChar
                        String modified = s.substring(0, i) + newChar + s.substring(i + 1);
                        maxPartitions = Math.max(maxPartitions, getMaxPartitions.applyAsInt(modified));
                    }
                }
            }
        }

        return maxPartitions;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.maxPartitionsAfterOperations("accca", 2)); // Example 1
        System.out.println(sol.maxPartitionsAfterOperations("aabaab", 3)); // Example 2
        System.out.println(sol.maxPartitionsAfterOperations("xxyz", 1)); // Example 3
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3003_maximize-the-number-of-partitions-after-operations/solutions/python3/solution.py
class Solution:
    def maxPartitionsAfterOperations(self, s: str, k: int) -> int:
        from collections import defaultdict
        
        def max_partitions(s: str) -> int:
            partitions = 0
            i = 0
            while i < len(s):
                count = defaultdict(int)
                distinct = 0
                j = i
                while j < len(s) and (distinct < k or (distinct == k and s[j] in count)):
                    if count[s[j]] == 0:
                        distinct += 1
                    count[s[j]] += 1
                    if distinct > k:
                        break
                    j += 1
                partitions += 1
                i = j
            return partitions
        
        # Check without any modification
        max_partitions_result = max_partitions(s)
        
        if k == 1:
            # If k == 1, each character change potentially maximizes partitions
            for i in range(len(s)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    if s[i] != c:
                        new_s = s[:i] + c + s[i+1:]
                        max_partitions_result = max(max_partitions_result, max_partitions(new_s))
            return max_partitions_result
        
        # Check by modifying each character once and replace it with every possible character
        for i in range(len(s)):
            original_char = s[i]
            for c in 'abcdefghijklmnopqrstuvwxyz':
                if c != original_char:
                    new_s = s[:i] + c + s[i+1:]
                    max_partitions_result = max(max_partitions_result, max_partitions(new_s))
        
        return max_partitions_result
